samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java

```diff
@@ -217,12 +250,28 @@ public void stop() {
       Thread.currentThread().interrupt();
     }
 
-    if (metrics != null) {
+    if (this.diagnosticsManager.isDefined()) {
+      try {
+        this.diagnosticsManager.get().stop();
+      } catch (InterruptedException e) {
+        log.error("InterruptedException while stopping diagnosticsManager", e);
+      }
+    }
+
+    if (containerProcessManagerMetrics != null) {
       try {
-        metrics.stop();
-        log.info("Stopped metrics reporters");
+
+        if (this.metricsReporters != null) {
```

Comment:
Should this be independent of whether CPMMetrics are defined or not?

Reply:
There are numerous tests that fail if i dont have this.

Reply:
I mean this block can be outside of the `if (containerProcessManagerMetrics != null)` check by itself.